daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/PrimitivesTextNumber.scala

```diff
@@ -275,17 +410,67 @@ case class ConvertTextStandardNumberPrim(e: ElementBase)
+    if (hasV) {
+      val r = TextNumberPatternUtils.vRegexStandard
+      r.findFirstMatchIn(patternNoQuoted) match {
+        case Some(r(_, _, afterV, _, _, _, _, _)) => afterV.length
+        case None =>
+          e.SDE(
+            s"""The dfdl:textNumberPattern '%s' contains 'V' (virtual decimal point).
+               | Other than the sign indicators, it can contain only
+               | '#', then digits 0-9 then 'V' then digits 0-9.
+               | The positive part of the dfdl:textNumberPattern is mandatory.""".stripMargin('|'),
+            pattern)
+      }
+    } else if (hasP) {
+      val rr = TextNumberPatternUtils.pOnRightRegexStandard
+      val rl = TextNumberPatternUtils.pOnLeftRegexStandard
+      val rightMatch = rr.findFirstMatchIn(patternNoQuoted)
+      val leftMatch = rl.findFirstMatchIn(patternNoQuoted)
+      (leftMatch, rightMatch) match {
+        case (None, None) => e.SDE(
+        """The dfdl:textNumberPattern '%s' contains 'P' (virtual decimal point positioners).
+            |However, it did not match the allowed syntax which allows the sign indicator
+            |plus digits on only one side of the P symbols.""".stripMargin,
           pattern)
+        case (Some(rl(_, ps, digits, _, negPre, negPs, negDigits, _)), None) => {
+          checkPPPMatchSyntax(ps, digits, negPre, negPs, negDigits)
+          ps.length + digits.length
+        }
+        case (None, Some(rr(_, digits, ps, _, negPre, negDigits, negPs, _))) => {
+          checkPPPMatchSyntax(ps, digits, negPre, negPs, negDigits)
+          - ps.length // negate value.
+        }
+        case _ => Assert.invariantFailed("Should not match both left P and right P regular expressions.")
+      }
+    } else {
+      0 // there is no V nor P, so there is no virtual decimal point scaling involved.
+    }
+  }
+
+  /**
+   * Check for common errors in the pattern matches for 'P' patterns.
+   *
+   * @param ps        match string for positive pattern P characters
+   * @param digits    match string for positive pattern digits (and sharps)
+   * @param negPre    negative prefix match (null if there is no negative pattern at all)
+   * @param negPs     match string for negative pattern P character (or null)
+   * @param negDigits match string for negative pattern P digits (and sharps)
+   */
+  private def checkPPPMatchSyntax(ps: String, digits: String, negPre: String, negPs: String, negDigits: String): Unit = {
+    Assert.invariant(ps.length >= 1)
+    digits.length >= 1
+    if (negPre ne null) {
+      e.schemaDefinitionUnless(ps == negPs,
+        """In the dfdl:textNumberPattern '%s' the P Symbols do not match
+          | between positive and negative patterns""".stripMargin,
+        pattern)
+      e.schemaDefinitionUnless(digits == negDigits,
+        """In the dfdl:textNumberPattern '%s' the '#' and digit symbols do not match
+          | between positive and negative patterns""".stripMargin,
+        pattern)
     }
```

Comment:
Looking back at the spec I saw this line:

> If there is an explicit negative subpattern, it serves only to specify the negative prefix and suffix; the number of digits, minimal digits, and other characteristics are ignored in the negative subpattern

Which gives me two questions:
1. Should we allow the negative and positive P's to be different? Seems like it's probably a user doing something they didn't intend if that's the case so maybe it's is worth an SDE as you have. Or at the very least this should be a warning.
2. Similarly, should the V and P regex patterns be much more lax about what they allow in the negative part since it's ignored? Again, if they don't match it probably an accident and an SDE seems very reasonable, but maybe there are schemas that will just want to keep the textNumberPattern as simple as possible and do something like `+000V0000;-0` so that they can specific the virtual decimal point in the postive pattern, and then just the prefix/suffix in the negative pattern and not worry about them having to match.